agentendpoint/config_task.go

```diff
@@ -0,0 +1,462 @@
+			ApplyConfigTaskProgress: &agentendpointpb.ApplyConfigTaskProgress{State: configState},
+		},
+	}
+	res, err := c.client.reportTaskProgress(ctx, req)
+	if err != nil {
+		return fmt.Errorf("error reporting task progress %s: %v", configState, err)
+	}
+	if res.GetTaskDirective() == agentendpointpb.TaskDirective_STOP {
+		return errServerCancel
+	}
+	return nil
+}
+
+func (c *configTask) validation(ctx context.Context) {
+	// First populate validation results and internal assignment state.
+	c.assignments = map[string]*assignment{}
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		asgnmnt := &assignment{policies: map[string]*policy{}}
+		c.assignments[a.GetConfigAssignment()] = asgnmnt
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := &policy{resources: map[string]*resource{}}
+			asgnmnt.policies[p.GetId()] = plcy
+			pResult := aResult.GetOsPolicyResults().GetResults()[i]
+			for i, r := range p.GetResources() {
+				plcy.resources[r.GetId()] = &resource{}
+				result := pResult.GetResourceResults().GetResults()[i]
+				result.GetExecutionSteps()[validationStepIndex] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep{
+					Step: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep_Validation{
+						Validation: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation{
+							Outcome: agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation_SKIPPED,
+						},
+					},
+				}
+			}
+		}
+	}
+
+	// Unmarshal all resources.
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		ctx = clog.WithLabels(ctx, map[string]string{"config_assignment": a.GetConfigAssignment()})
+		asgnmnt := c.assignments[a.GetConfigAssignment()]
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := asgnmnt.policies[p.GetId()]
+			pResult := aResult.GetOsPolicyResults().GetResults()[i]
+			for i, r := range p.GetResources() {
+				res := plcy.resources[r.GetId()]
+
+				outcome := agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation_OK
+				errMsg := ""
+				err := res.unmarshal(ctx, r)
+				if err != nil {
+					outcome = agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation_RESOURCE_PAYLOAD_CONVERSION_ERROR
+					plcy.hasError = true
+					errMsg = fmt.Sprintf("Error unmarshalling resource: %v", err)
+					clog.Errorf(ctx, errMsg)
+				}
+
+				result := pResult.GetResourceResults().GetResults()[i]
+				result.GetExecutionSteps()[validationStepIndex] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep{
+					Step: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep_Validation{
+						Validation: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation{
+							Outcome: outcome,
+						},
+					},
+					ErrMsg: errMsg,
+				}
+
+				// Should we continue to validate all resources, or stop on the first unmarshal error?
```

Comment:
If it's all the same, reporting an error for all the resources would be more helpful to users.

Reply:
That was my thinking as well

Reply:
We should continue to validate other resources so that all resource errors can be reported together and fixed.